ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/DecryptPioneerPayloads.java

```diff
@@ -81,11 +91,36 @@ private DecryptPioneerPayloads(ValueProvider<String> metadataLocation,
         keyStore = KeyStore.of(metadataLocation.get(), kmsEnabled.get());
       }
 
+      if (validator == null) {
+        validator = new JsonValidator();
+        byte[] data = Resources
+            .toByteArray(Resources.getResource("telemetry.pioneer-study.4.schema.json"));
+        envelopeSchema = JSONSchemaStore.readSchema(data);
+      }
+
       ObjectNode json = Json.readObjectNode(message.getPayload());
-      PrivateKey key = keyStore.getKey(message.getAttribute(Attribute.DOCUMENT_NAMESPACE));
+      validator.validate(envelopeSchema, json);
+      JsonNode payload = json.get("payload");
+
+      String encryptionKeyId = payload.get("encryptionKeyId").asText();
+      PrivateKey key = keyStore.getKey(encryptionKeyId);
+      if (key == null) {
+        // Is this really an IOException?
+        throw new IOException(String.format("encryptionKeyId not found: %s", encryptionKeyId));
+      }
+
+      // NOTE: there could be a method for handling merging metadata like
```

Comment:
I'm not entirely sure what need is being described here, but see `AddMetadata.java` where we insert metadata at the beginning of a json payload without parsing it.

Reply:
There's the `application` field at the top level part of the ping (as part of the common ping format). It could be appended to the actual payload so it can be available to analysis downstream. The problem noted here is that the payload could be compressed before encryption, which would make it difficult to concatenate unless it were decompressed or somehow done in the post-decompression step.